File: 98. Validate Binary Search Tree/solution2.py
from typing import *

class TreeNode:
    """Definition for a binary tree node."""
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def valid(node, minimum, maximum):
            if not node:
                return True
            if not (minimum < node.val < maximum):
                return False
            
            return valid(node.left, minimum, node.val) and valid(node.right, node.val, maximum)

        return valid(root, float('-inf'), float('inf'))

File: 98. Validate Binary Search Tree/solution2.js
/** Definition for a binary tree node. */
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
    function valid(node, minimum, maximum) {
        if (!node) {
            return true;
        }

        if (node.val <= minimum || node.val >= maximum) {
            return false;
        }

        return valid(node.left, minimum, node.val) && valid(node.right, node.val, maximum);
    }

    return valid(root, -Infinity, Infinity);
};